dask/order.py

```diff
@@ -115,787 +106,462 @@ def order(
-    current: list[Key] = []
-    current_pop = current.pop
-    current_append = current.append
-    for key, deps in dependents.items():
-        if not deps:
-            val = total_dependencies[key]
-            result[key] = (1, val, val, 0, 0)
-            for child in dependencies[key]:
+    current = []
+    num_needed = {k: len(v) for k, v in dependencies.items() if v}
+    max_dependents = {}
+    for k, v in dependencies.items():
+        if not v:
+            result[k] = {k}
+            deps = dependents[k]
+            max_dependents[k] = len(deps)
+            for child in deps:
                 num_needed[child] -= 1
                 if not num_needed[child]:
-                    current_append(child)
-
+                    current.append(child)
     while current:
-        key = current_pop()
-        parents = dependents[key]
-        if len(parents) == 1:
-            (parent,) = parents
-            (
-                total_dependents,
-                min_dependencies,
-                max_dependencies,
-                min_heights,
-                max_heights,
-            ) = result[parent]
-            result[key] = (
-                1 + total_dependents,
-                min_dependencies,
-                max_dependencies,
-                1 + min_heights,
-                1 + max_heights,
-            )
-        else:
-            (
-                total_dependents_,
-                min_dependencies_,
-                max_dependencies_,
-                min_heights_,
-                max_heights_,
-            ) = zip(*(result[parent] for parent in dependents[key]))
-            result[key] = (
-                1 + sum(total_dependents_),
-                min(min_dependencies_),
-                max(max_dependencies_),
-                1 + min(min_heights_),
-                1 + max(max_heights_),
-            )
-        for child in dependencies[key]:
+        key = current.pop()
+        for child in dependents[key]:
             num_needed[child] -= 1
             if not num_needed[child]:
-                current_append(child)
-    return result
+                current.append(child)
+        # At some point, all the roots are the same, particualarly for dense
+        # graphs. We don't want to create new sets over and over again
+        new_set = set()
+        previous: set[Key] = set()
+        identical_sets = True
+        for parent in dependencies[key]:
+            if not previous:
+                previous = result[parent]
+                max_dependents[key] = max_dependents[parent]
+            elif identical_sets and (
+                previous is result[parent]
+                or (len(previous) == len(result[parent]) and previous == result[parent])
+            ):
+                identical_sets = True
+            else:
+                identical_sets = False
+                max_dependents[key] = max(max_dependents[parent], max_dependents[key])
+                new_set.update(result[parent])
```

Comment:
yeah, this evolved over time and I rewrote it already once or twice. I'm fine with your suggestions

Reply:
logically, those expressions are identical but they do observe different performance

consider the case where `identical_sets=True` and `previous is result[parent]` 

My former implementation would evaluate to `not True or True or <expensive-check>` where `<expensive-check>` is the length plus content check, i.e. for the exact same python set/object we would skip the more expensive check for equal elements.

Your implementation otoh would evaluate to

`False or (False and <expensive-check>)`

i.e. we could not benefit from short circuiting. I'll revert this again

Reply:
however, I also notice a bug so I'll rewrite this again

Reply:
> Your implementation otoh would evaluate to
> 
> `False or (False and <expensive-check>)`
> 
> i.e. we could not benefit from short circuiting. I'll revert this again

FWIW, `False or (False and <expensive_check>)` also short-circuits because `False and <whatever>` is always `False`.